### ISSUE TITLE
Remove discovery plug-in in favor of hitting Service

### ISSUE THREAD

**Issue by ant31** [pires/docker-elasticsearch-kubernetes#3]
Hi,

I successfully try to remove Elasticsearch discovery plugin Kubernetes

The new configuration is simple:
```yaml
discovery:
  zen.ping.unicast.hosts: ${UNICAST_HOSTS}
```

and in the  ES-RC: 
```yaml
 env:
        - name: UNICAST_HOSTS
          value: 'es-discovery.default.svc.cluster.local'
```

Is this plugin doing something more than discovering new nodes?

**Comment by jimmidyson** [comment]
No this plugin doesn't do any more than that. What happens when you try to disable?

**Comment by ant31** [comment]
From what I ve observed, when I scale or start the cluster  elasticsearch is trying to find another es-node by using the elasticsearch-discovery service (service for masters only on 9300), so it find a master and register itself into the cluster.

**Comment by jimmidyson** [comment]
If you want to _only_ use unicast then I think you need to disable multicast explicitly - https://www.elastic.co/guide/en/elasticsearch/guide/current/_important_configuration_changes.html#CO309-1

**Comment by pires** [comment]
Yes @ant31 this is possible, yet I'm more comfortable with having all nodes connecting to all masters available for discovery. Anyway, open to discussion.

**Comment by dermidgen** [comment]
Okay, so along these lines - I'm not stuck when using `kubectl scale --replicas=3 rc es`.  I have fabric8 enabled, I'm using the default image here so multicast is set to false. I'm hitting:

```
[2015-11-20 07:14:46,829][INFO ][node                     ] [Trapper] initialized
[2015-11-20 07:14:46,832][INFO ][node                     ] [Trapper] starting ...
[2015-11-20 07:14:47,064][WARN ][common.network           ] [Trapper] publish address: {0.0.0.0} is a wildcard address, falling back to first non-loopback: {10.56.0.12}
[2015-11-20 07:14:47,065][INFO ][transport                ] [Trapper] publish_address {10.56.0.12:9300}, bound_addresses {[::]:9300}
[2015-11-20 07:14:47,096][INFO ][discovery                ] [Trapper] myesdb/nZLj16NNTpmWh9e1z153hA
[2015-11-20 07:14:47,099][WARN ][common.network           ] [Trapper] publish address: {0.0.0.0} is a wildcard address, falling back to first non-loopback: {10.56.0.12}
[2015-11-20 07:14:48,918][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9200
[2015-11-20 07:14:48,918][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9300
[2015-11-20 07:14:50,448][WARN ][common.network           ] [Trapper] publish address: {0.0.0.0} is a wildcard address, falling back to first non-loopback: {10.56.0.12}
[2015-11-20 07:14:50,458][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9200
[2015-11-20 07:14:50,458][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9300
[2015-11-20 07:14:51,974][WARN ][common.network           ] [Trapper] publish address: {0.0.0.0} is a wildcard address, falling back to first non-loopback: {10.56.0.12}
[2015-11-20 07:14:51,978][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9200
[2015-11-20 07:14:51,979][INFO ][io.fabric8.elasticsearch.discovery.kubernetes.KubernetesUnicastHostsProvider] [Trapper] adding endpoint /10.56.1.6, transport_address 10.56.1.6:9300
[2015-11-20 07:14:52,018][INFO ][cluster.service          ] [Trapper] new_master {Trapper}{nZLj16NNTpmWh9e1z153hA}{10.56.0.12}{10.56.0.12:9300}{master=true}, reason: zen-disco-join(elected_as_master, [0] joins received)
[2015-11-20 07:14:52,051][WARN ][common.network           ] [Trapper] publish address: {0.0.0.0} is a wildcard address, falling back to first non-loopback: {10.56.0.12}
[2015-11-20 07:14:52,051][INFO ][http                     ] [Trapper] publish_address {10.56.0.12:9200}, bound_addresses {[::]:9200}
[2015-11-20 07:14:52,056][INFO ][node                     ] [Trapper] started
[2015-11-20 07:14:52,062][INFO ][gateway                  ] [Trapper] recovered [0] indices into cluster_state
[2015-11-20 07:15:03,002][WARN ][http.netty               ] [Trapper] Caught exception while handling client http traffic, closing connection [id: 0x7bb01c3e, /10.56.2.6:53400 => /10.56.0.12:9200]
java.lang.IllegalArgumentException: invalid version format: INTERNAL:DISCOVERY/ZEN/UNICASTﾲ￐^MYESDBANGELICA JONESY-CPTOLCTSAR5UFH3XEHPG	10.56.2.6	10.56.2.6
	at org.jboss.netty.handler.codec.http.HttpVersion.<init>(HttpVersion.java:94)
	at org.jboss.netty.handler.codec.http.HttpVersion.valueOf(HttpVersion.java:62)
	at org.jboss.netty.handler.codec.http.HttpRequestDecoder.createMessage(HttpRequestDecoder.java:75)
	at org.jboss.netty.handler.codec.http.HttpMessageDecoder.decode(HttpMessageDecoder.java:191)
	at org.jboss.netty.handler.codec.http.HttpMessageDecoder.decode(HttpMessageDecoder.java:102)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:500)
...
```

**Comment by jimmidyson** [comment]
tl;dr: that looks like you've set up your services wrongly - need separate services for http port (9200) & cluster comms (9300). Same error reported [here](https://github.com/fabric8io/elasticsearch-cloud-kubernetes/issues/18#issuecomment-158155523) with discussion [here](https://github.com/fabric8io/elasticsearch-cloud-kubernetes/issues/18#issuecomment-158176679).

**Comment by pires** [comment]
Test with https://github.com/pires/kubernetes-elasticsearch-cluster/pull/25 and report back please. I tested it with 4 nodes, both Kubernetes 1.1.1 and GKE, and it worked smoothly.

**Comment by ant31** [comment]
@pires what have you tested? 

> I'm more comfortable with having all nodes connecting
I think as soon a node is part of the cluster if its initial master goes down it won't affect it.

**Comment by jimmidyson** [comment]
@ant31 Ah I think I understand what your trying to do now. You're trying to use a single service to find the master without the discovery plugin? If that works great then that's brilliant & we have no need for the discovery plugin at all. Definitely worth testing out.

**Comment by jimmidyson** [comment]
Well this seems to work perfectly! How simple! I'm sure this never used to work, it was my first idea to do it this way, it didn't work so I wrote the plugin. But I'm really happy to see it working now!

@pires Would you be able to test it out too to confirm?

**Comment by pires** [comment]
@jimmidyson I understood what @ant31 is proposing and it's really neat given its simplicity. But again, I feel more comfortable with a node hitting all masters. Distributed systems are tricky and we can never rely on a single master being available - a service may point to a unhealthy pod that's not yet (or ever) marked as unhealthy - to join a cluster.

**Comment by jimmidyson** [comment]
@pires It does actually hit all masters... well it hits the service which routes to any endpoint which in turn returns all nodes in the cluster. It then makes peer-to-peer connections to masters.

**Comment by pires** [comment]
@jimmidyson my knowledge on Services is that it routes a connection request to a single pod. Have I been wrong all this time?

**Comment by pires** [comment]
I'd like to have input from the Kubernetes people (@erictune @brendanburns @kelseyhightower) because there are other example apps that would also work with this simple trick, like Hazelcast and Cassandra.

**Comment by ant31** [comment]
@pires, yes but hitting the service is its exactly like having the full pod list (in the case of ES)
It will route to one available master and then the node receive complete cluster view, it's never bound to a single point.

**Comment by pires** [comment]
@ant31 what happens if the node to which the Service points is malfunctioning (and assuming there's no liveness probe implemented)? What happens if the connection fails? To my understanding, ES will not retry discovery but even if it does it's not guaranteed it won't reach the same borked master node.

**Comment by ant31** [comment]
@pires, good point. Using the service liveness/readiness probe would be a requirement(is that a matter?) 

To argument your way I've just found an issue:  
scaling master quickly (like from 1 to 5) without probe: new master have a high chance to cluster itself with another new master, and then not being part of the targeted cluster.

**Comment by jimmidyson** [comment]
I wonder if this is any different than using the discovery plugin?

**Comment by pires** [comment]
@jimmidyson yes, because the plug-in hits _all_ the nodes we tell it to - in my case and the Kubernetes _official_ Elasticsearch example, all master nodes.

**Comment by pires** [comment]
@ant31 unfortunately, Elasticsearch does not share cluster topology (gossip), like Hazelcast and Cassandra do.

**Comment by jimmidyson** [comment]
Now that is a very good point. OK you've convinced me :)

**Comment by dermidgen** [comment]
@jimmidyson I did see that thread and the issue there. What I am trying to do is literally follow the guide https://github.com/kubernetes/kubernetes/tree/master/examples/elasticsearch and figure out what updates I need to make for ES 2.  At this point, literally following that guide without deviation, got me that result. My guess is because of the single mixed service instance being scaled with `kubectl scale`.

I'm thinking maybe I'm in the wrong place for this; really I should just be updating the .yaml files in the official docs, right?

**Comment by jimmidyson** [comment]
I think @pires will be doing that soon.

**Comment by pires** [comment]
@dermidgen that tutorial is not compatible with 1.1.x. Again, the changes in https://github.com/pires/kubernetes-elasticsearch-cluster/pull/25 are enough.

**Comment by pires** [comment]
@jimmidyson since I ignored one PR by @dermidgen I'm asking him to do it for me. It's a win/win scenario :D

**Comment by ant31** [comment]
> Elasticsearch does not share cluster topology 

Yes the cluster has to have at least one master, but each nodes knows all other node as soon as they join/left. 

I've tried following scenario:
-  1 master / 2 clients, then remove the master  0 master / 2 clients 
-  scaled up to 1 master, the cluster was back healthy, 
-  scaled up to 2 master, the cluster still healthy. 
-  remove first master, the cluster still healthy

I think the discovery is working as expected, and none of the node is bound to one pod. 
The case it tries to cluster against an unhealthy master is still available. But as soon as a es-node joined the cluster it act the same way with or without the plugin.

Here some logs from elasticsearch:
 
a - 
1 master (Killer Shrike).
scale es-clients from 0 to 2:
Master log:
```
[2015-11-20 16:01:39,730][INFO ][cluster.service          ] [Killer Shrike] added {{Phage}{CcvNpESxQ0-yzSzIM18E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, \
master=false},}, reason: zen-disco-join(join from node[{Phage}{CcvNpESxQ0-yzSzIM18E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, master=false}])
[2015-11-20 16:01:39,958][INFO ][cluster.service          ] [Killer Shrike] added {{Candra}{xri1_zYzTCm3fKauk7tiTg}{192.168.0.6}{192.168.0.6:9300}{data=false,\
 master=false},}, reason: zen-disco-join(join from node[{Candra}{xri1_zYzTCm3fKauk7tiTg}{192.168.0.6}{192.168.0.6:9300}{data=false, master=false}])
```

b-
 Scale master to 0
client logs:
```
2. [2015-11-20 16:04:02,355][INFO ][discovery.zen            ] [Phage] master_left [{Killer Shrike}{mw5zw-IsQUauOpoZtMsyBg}{192.168.0.13}{192.168.0.13:9300}{d\
ata=false, master=true}], reason [transport disconnected]
[2015-11-20 16:04:02,356][WARN ][discovery.zen            ] [Phage] master left (reason = transport disconnected), current nodes: {{Phage}{CcvNpESxQ0-yzSzIM18\
E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, master=false},{Candra}{xri1_zYzTCm3fKauk7tiTg}{192.168.0.6}{192.168.0.6:9300}{data=false, master=false},}
[2015-11-20 16:04:02,357][INFO ][cluster.service          ] [Phage] removed {{Killer Shrike}{mw5zw-IsQUauOpoZtMsyBg}{192.168.0.13}{192.168.0.13:9300}{data=fal\
se, master=true},}, reason: zen-disco-master_failed ({Killer Shrike}{mw5zw-IsQUauOpoZtMsyBg}{192.168.0.13}{192.168.0.13:9300}{data=false, master=true})
[2015-11-20 16:04:19,420][DEBUG][action.admin.cluster.health] [Phage] no known master node, scheduling a retry
```
Client are in errors, 'no known master'

c- 
Scale master back to 1:
master  logs:
```
[2015-11-20 16:05:35,441][INFO ][cluster.service          ] [Crossfire] added {{Candra}{xri1_zYzTCm3fKauk7tiTg}{192.168.0.6}{192.168.0.6:9300}{data=false, mas\
ter=false},}, reason: zen-disco-join(join from node[{Candra}{xri1_zYzTCm3fKauk7tiTg}{192.168.0.6}{192.168.0.6:9300}{data=false, master=false}])
[2015-11-20 16:05:35,463][INFO ][cluster.service          ] [Crossfire] added {{Phage}{CcvNpESxQ0-yzSzIM18E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, mast\
er=false},}, reason: zen-disco-join(join from node[{Phage}{CcvNpESxQ0-yzSzIM18E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, master=false}])
```
As you see the new master immediatly knows about the two existing client

Same for clients, they immediatly get in touch with the new master
clients logs:
```
[2015-11-20 16:05:35,479][INFO ][cluster.service          ] [Phage] detected_master {Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=\
false, master=true}, added {{Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=false, master=true},}, reason: zen-disco-receive(from ma\
ster [{Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=false, master=true}])

[2015-11-20 16:05:35,454][INFO ][cluster.service          ] [Candra] master {new {Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=fal\
se, master=true}}, removed {{Phage}{CcvNpESxQ0-yzSzIM18E4w}{192.168.0.5}{192.168.0.5:9300}{data=false, master=false},}, added {{Crossfire}{-wJ03RW2R7WbRBjKFc4\``
djA}{192.168.0.11}{192.168.0.11:9300}{data=false, master=true},}, reason: zen-disco-receive(from master [{Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192\
.168.0.11:9300}{data=false, master=true}])

```
The cluster is working again

d- 
Now I tryed to scale to 2 and remove the Crossfire master, to check if its bound to the first found.. 

scale master to 2:
client:
```
[2015-11-20 16:08:10,934][INFO ][cluster.service          ] [Candra] added {{Ryder}{e2WZCZCPQ2mGtXolxi4kqA}{192.168.0.12}{192.168.0.12:9300}{data=false, maste\
r=true},}, reason: zen-disco-receive(from master [{Crossfire}{-wJ03RW2R7WbR``BjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=false, master=true}])
```

downscale scale master to 1:
client:
```
[2015-11-20 16:09:46,915][INFO ][cluster.service          ] [Candra] removed {{Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=false,\
 master=true},}, reason: zen-disco-master_failed ({Crossfire}{-wJ03RW2R7WbRBjKFc4djA}{192.168.0.11}{192.168.0.11:9300}{data=false, master=true})
``

[2015-11-20 16:09:51,453][INFO ][cluster.service          ] [Candra] detected_master {Ryder}{e2WZCZCPQ2mGtXolxi4kqA}{192.168.0.12}{192.168.0.12:9300}{data=fal\
se, master=true}, reason: zen-disco-receive(from master [{Ryder}{e2WZCZCPQ2mGtXolxi4kqA}{192.168.0.12}{192.168.0.12:9300}{data=false, master=true}])
```

The clients correctly detected `removed {{Crossfire}``, and added the new elected master.


If you want I can try another scenario and share results.

**Comment by dermidgen** [comment]
@jimmidyson @pires I've opened a PR here: https://github.com/kubernetes/kubernetes/pull/17570

There's a couple outstanding questions I still have related to the simple example and `emptyDir` warnings. Love it if you guys could weigh in there.

**Comment by jimmidyson** [comment]
I think it might be worth asking the elasticsearch team if this approach is recommended? Discovery via a load balancer. Hmm. It seems to work really well but there might be some nuance we're missing. I certainly hope it is supported, that means one less project for me to maintain ;)

**Comment by jimmidyson** [comment]
Asked [here]( https://discuss.elastic.co/t/elasticsearch-2-0-unicast-discovery-via-load-balancer/35210)

**Comment by jimmidyson** [comment]
Good reply [here](https://discuss.elastic.co/t/elasticsearch-2-0-unicast-discovery-via-load-balancer/35210/5?u=jimmidyson) suggesting we should continue with discovery via container IPs, but that the headless cluster service should only point to masters, not any node in the cluster - this is just a deployment change (selector on cluster service), not code change. SGTM:

> The unicast hosts used to exchange information between new nodes of the cluster. If the cluster is already formed, a new node can ping any node and ask it for the current master. However, if you have a full cluster restart, where all the new nodes are pinging and discovering each other, a LB can (in theory) lead them to miss nodes. For example, if consistently separates the nodes into two halves, those two halves will never discover each other. Since you have mentioned using dedicated master nodes, I would suggest just putting those in the list. Last, since you talk about kubernetes, maybe it has an API which can give the list of current container IPs? If so you can write a little discovery plugin to ingest that list as a unicast host seed (see how the GCE/EC2 plugins work now).

**Comment by pires** [comment]
@jimmidyson thanks a lot.

* [My service for discovery purposes alone](https://github.com/pires/kubernetes-elasticsearch-cluster/blob/master/es-discovery-svc.yaml)
* [My service for client usage alone](https://github.com/pires/kubernetes-elasticsearch-cluster/blob/master/es-svc.yaml)

**Comment by jimmidyson** [comment]
Perfect!

**Comment by ant31** [comment]
> that means one less project for me to maintain ;)
Not this time :)

Thank you @pires, @jimmidyson  to have investigate on this.

**Comment by pires** [comment]
@ant31 thank you for reporting and keeping with us.

**Comment by jimmidyson** [comment]
@ant31 Yes thank for the idea - got me excited for a little while ;)
